copy/copy.go

```diff
@@ -472,19 +472,33 @@ func (ic *imageCopier) copyBlobFromStream(srcStream io.Reader, srcInfo types.Blo
 
 	// === Compress the layer if it is uncompressed and compression is desired
 	var inputInfo types.BlobInfo
-	if !canCompress || isCompressed || !ic.dest.ShouldCompressLayers() {
+	if !canCompress || isCompressed == ic.dest.ShouldCompressLayers() {
 		logrus.Debugf("Using original blob without modification")
 		inputInfo = srcInfo
 	} else {
 		logrus.Debugf("Compressing blob on the fly")
-		pipeReader, pipeWriter := io.Pipe()
-		defer pipeReader.Close()
 
-		// If this fails while writing data, it will do pipeWriter.CloseWithError(); if it fails otherwise,
-		// e.g. because we have exited and due to pipeReader.Close() above further writing to the pipe has failed,
-		// we don’t care.
-		go compressGoroutine(pipeWriter, destStream) // Closes pipeWriter
-		destStream = pipeReader
+		// Decompress the original stream if it is compressed.
+		if isCompressed {
+			logrus.Debugf("... but first decompressing the original blob")
+			destStream, err = decompressor(destStream)
+			if err != nil {
+				return types.BlobInfo{}, errors.Wrapf(err, "Error decompressing blob %s on-the-fly", srcInfo.Digest)
+			}
+		}
```

Comment:
Note that we already might be decompressing the stream once, in the `getOriginalLayerCopyWriter` path above, and doing the same work twice is rather wasteful. OTOH integrating all of that here _is_ close to madness.

(Eventually the `imageCopier` needs to grow good tests. Not today, I guess :) )